vent/coordinator/coordinator.py

```diff
@@ -0,0 +1,156 @@
+import time
+import threading
+from typing import List
+
+from vent.common.message import SensorValues, ControlSetting, Alarm, ControlSettingName, IPCMessageCommand
+from vent.controller.control_module import get_control_module
+from vent.coordinator.ipc import IPC
+from vent.coordinator.process_manager import ProcessManager
+
+
+class CoordinatorBase:
+    def __init__(self, sim_mode=False):
+        # get_ui_control_module handles single_process flag
+        self.control_module = get_control_module(sim_mode)
+        self.sensor_values = None
+        self.alarms = None
+        self.control_settings = {}
+        self.tentative_control_settings = {}
+        self.last_message_timestamp = None
+
+    def get_sensors(self) -> SensorValues:
+        # returns SensorValues struct
+        pass
+
+    # def get_alarms(self) -> List[Alarm]:
+    #     # returns list of Alarm structs
+    #     pass
+
+    def get_active_alarms(self) -> List[Alarm]:
+        pass
+
+    def get_logged_alarms(self) -> List[Alarm]:
+        pass
+
+    def clear_logged_alarms(self):
+        pass
+
+    def set_control(self, control_setting: ControlSetting):
+        # takes ControlSetting struct
+        pass
+
+    def get_control(self, control_setting_name: ControlSettingName) -> ControlSetting:
+        pass
+
+    def get_msg_timestamp(self):
+        # return timestamp of last message
+        pass
+
+    def do_process(self, command: IPCMessageCommand):
+        # TODO: we need to test these
+        # start / stop / reset process
+        if command == IPCMessageCommand.START:
+            self.control_module.start()
+        elif command == IPCMessageCommand.STOP:
+            self.control_module.stop()
+        else:
+            raise KeyError("Error: undefined action" + str(command))
+
+
+class CoordinatorLocal(CoordinatorBase):
+    def __init__(self, sim_mode=False):
+        super().__init__(sim_mode=sim_mode)
+        self.lock = threading.Lock()
+        self.thread = threading.Thread(target=self.start, daemon=True)
+        self.thread.start()
+        self.thread_id = self.thread.ident
+
+    def __del__(self):
+        self.thread.join()
+
+    def set_control(self, control_setting: ControlSetting):
+        self.lock.acquire()
+        self.tentative_control_settings[control_setting.name] = control_setting
+        self.lock.release()
+
+    def get_control(self, control_setting_name: ControlSettingName) -> ControlSetting:
+        self.lock.acquire()
+        control_setting = self.control_settings[control_setting_name]
+        self.lock.release()
+        return control_setting
+
+    def get_sensors(self) -> SensorValues:
+        self.lock.acquire()
+        sensor_values = self.sensor_values
+        self.lock.release()
+        return sensor_values
+
+    def get_logged_alarms(self) -> List[Alarm]:
+        return self.control_module.get_logged_alarms()
```

Comment:
We should make a copy of the alarm list, or maybe that should be done in the controller. But either way we don't want to depend on the caller to take mutexes when using the alarms.

Reply:
Do you mean "we don't want to depend on the callee to take mutexes"? I think here I should keep a copy of the alarm, similar to the implementation in get_sensors/get_controls. What do you think?

By reading the code in `ControlModuleSimulator.update_alarm`, I think it is easier for the controller to trigger the alarm because it knows startpoint/endpoint of each waveform. The coordinator then will just keep a local copy of it. 

And another thing, @sneakers-the-rat how do you plan to clear the active alarm? I don't think coordinator right now has such API. Maybe you can define?

Reply:
Yes, keep a copy of the alarms sounds good. But we also need to return a copy of the copy to the UI caller so the UI doesn't have to worry about taking a mutex when it uses the alarm values.

Reply:
I can do that. The UI can assume the function return is a copy instead of a reference.